File: LeetCode/cpp/130.cpp
class Solution {
    vector<int> dir_x = {-1, 0, 1, 0}, dir_y = {0, -1, 0, 1};
    int m, n;
public:
    void solve(vector<vector<char>> &board) {
        m = board.size(), n = m == 0 ? 0 : board[0].size();
        if (m == 0)
            return;
        for (int i = 0; i < m; ++i) {
            if (board[i][0] == 'O')
                DFS(board, i, 0);
            if (board[i][n - 1] == 'O')
                DFS(board, i, n - 1);
        }
        for (int j = 1; j < n - 1; ++j) {
            if (board[0][j] == 'O')
                DFS(board, 0, j);
            if (board[m - 1][j] == 'O')
                DFS(board, m - 1, j);
        }
        for (int i = 0; i < m; ++i) {
            for (int j = 0; j < n; ++j) {
                auto &c = board[i][j];
                if (c == 'O')
                    c = 'X';
                else if (c == 'Y')
                    c = 'O';
            }
        }
    }

    void DFS(vector<vector<char>> &board, int i, int j) {
        board[i][j] = 'Y';
        for (int k = 0; k < 4; ++k) {
            int x = i + dir_x[k], y = j + dir_y[k];
            if (x >= 0 && x < m && y >= 0 && y < n && board[x][y] == 'O')
                DFS(board, x, y);
        }
    }
};

File: LeetCode/python3/130.py
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        def DFS(x, y):
            board[x][y] = 'M'
            dir_x, dir_y = [0, 1, 0, -1], [1, 0, -1, 0]
            for k in range(4):
                new_x, new_y = x + dir_x[k], y + dir_y[k]
                if new_x >= 0 and new_x < m and new_y >= 0 and new_y < n and board[new_x][new_y] == 'O':
                    DFS(new_x, new_y)

        m, n = len(board), len(board[0]) if len(board) > 0 else 0
        if m == 0:
            return
        for i in range(m):
            if board[i][0] == 'O':
                DFS(i, 0)
            if board[i][n - 1] == 'O':
                DFS(i, n - 1)
        for j in range(1, n - 1):
            if board[0][j] == 'O':
                DFS(0, j)
            if board[m - 1][j] == 'O':
                DFS(m - 1, j)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'M':
                    board[i][j] = 'O'
                elif board[i][j] == 'O':
                    board[i][j] = 'X'
           